pkg/util/pki/csr.go

```diff
@@ -278,148 +302,208 @@ func buildKeyUsagesExtensionsForCertificate(crt *v1.Certificate) ([]pkix.Extensi
-		rawSubject, err := ParseSubjectStringToRawDERBytes(crt.Spec.LiteralSubject)
-		if err != nil {
-			return nil, err
-		}
+func UpdateTemplateFromOptions(cert *x509.Certificate, duration time.Duration, isCA bool, maxPathLen *int, keyUsage x509.KeyUsage, extKeyUsage []x509.ExtKeyUsage) *x509.Certificate {
+	// Override certificate template values with values that are expected but
+	// not present in the CSR that generated the certificate template.
 
-		cert.RawSubject = rawSubject
+	cert.NotBefore = time.Now()
+	cert.NotAfter = cert.NotBefore.Add(duration)
+
+	// We expect that if the CSR contains a BasicConstraints extension, it
+	// matches the IsCA and MaxPathLen values that we set here.
+	// This should have been enforced by our validation webhook.
+
+	cert.BasicConstraintsValid = true
+	cert.IsCA = isCA
+	if maxPathLen != nil {
+		cert.MaxPathLen = *maxPathLen
+		cert.MaxPathLenZero = *maxPathLen == 0
 	} else {
-		cert.Subject = pkix.Name{
-			Country:            subject.Countries,
-			Organization:       organization,
-			OrganizationalUnit: subject.OrganizationalUnits,
-			Locality:           subject.Localities,
-			Province:           subject.Provinces,
-			StreetAddress:      subject.StreetAddresses,
-			PostalCode:         subject.PostalCodes,
-			SerialNumber:       subject.SerialNumber,
-			CommonName:         commonName,
+		cert.MaxPathLen = 0
+		cert.MaxPathLenZero = false
+	}
+
+	// We expect that if the CSR contains a KeyUsage/ ExtendedKeyUsage
+	// extension, these KeyUsages match the usages that we set here. To
+	// be safe, we append the usages instead of replacing them.
+	// This should have been enforced by our validation webhook.
+
+	cert.KeyUsage |= keyUsage
+OuterLoop:
+	for _, usage := range extKeyUsage {
+		for _, existingUsage := range cert.ExtKeyUsage {
+			if existingUsage == usage {
+				continue OuterLoop
+			}
 		}
+
+		cert.ExtKeyUsage = append(cert.ExtKeyUsage, usage)
 	}
 
-	return cert, nil
+	return cert
 }
 
-// GenerateTemplate will create a x509.Certificate for the given
-// CertificateRequest resource
-func GenerateTemplateFromCertificateRequest(cr *v1.CertificateRequest) (*x509.Certificate, error) {
-	certDuration := apiutil.DefaultCertDuration(cr.Spec.Duration)
-	keyUsage, extKeyUsage, err := KeyUsagesForCertificateOrCertificateRequest(cr.Spec.Usages, cr.Spec.IsCA)
+// GenerateTemplateFromCSRPEMWithUsages will create a x509.Certificate for the
+// given csrPEM.
+// Call UpdateTemplateFromOptions to override the duration, isCA, maxPathLen, keyUsage, and extKeyUsage.
+func GenerateTemplateFromCSRPEM(csrPEM []byte) (*x509.Certificate, error) {
+	csr, err := DecodeX509CertificateRequestBytes(csrPEM)
 	if err != nil {
 		return nil, err
 	}
-	return GenerateTemplateFromCSRPEMWithUsages(cr.Spec.Request, certDuration, cr.Spec.IsCA, keyUsage, extKeyUsage)
-}
-
-func GenerateTemplateFromCSRPEM(csrPEM []byte, duration time.Duration, isCA bool) (*x509.Certificate, error) {
```

Comment:
I removed this function, I think it should be much more explicit what values you use for some of the fields in your Certificate. In this old function we always set KeyUsages and ExtKeyUsages to 0/ nil.

Reply:
Also, I needed this name for a function with another function signature.

Reply:
Instead, the GenerateTemplateFromCSRPEMWithOptions function can be used.

Reply:
suggestion: I'm wary of making breaking changes to this package since it has quite a lot of [importers](https://pkg.go.dev/github.com/cert-manager/cert-manager@v1.11.1/pkg/util/pki?tab=importedby) in open source and presumably a few more in closed source.

We don't guarantee we won't break things, but it's nice to avoid it. I don't think re-using the name is worth breaking it - especially because it'll make it confusing when users upgrade only to see a changed function signature.

I'd vote to deprecate this function and pick a new name.

Reply:
@SgtCoDFish I now use the name `CertificateTemplateFromCSR` and kept the old `GenerateTemplateFromCSRPEM` function signature.